src/onnx/parse_reduce_op.cpp

```diff
@@ -31,66 +31,110 @@ namespace migraphx {
 namespace onnx {
 
-instruction_ref parse_reduce_oper(const std::string& op_name,
-                                  const onnx_parser& parser,
-                                  onnx_parser::node_info info,
-                                  std::vector<instruction_ref> args)
+template <typename Derived>
+struct reduce_parser : op_parser<Derived>
 {
-    // default to reduce over all dimensions
-    std::vector<int64_t> axes;
-    if(args.size() == 2)
+    instruction_ref parse_reduce_oper(const std::string& op_name,
+                                      const onnx_parser& parser,
+                                      onnx_parser::node_info info,
+                                      std::vector<instruction_ref> args) const
     {
-        auto arg_axes = args.at(1)->eval();
-        check_arg_empty(arg_axes, "PARSE_" + op_name + ": cannot handle variable axes!");
-        axes.clear();
-        arg_axes.visit([&](auto s) { axes.assign(s.begin(), s.end()); });
-    }
-    else if(contains(info.attributes, "axes"))
-    {
-        axes.clear();
-        auto&& attr_axes = info.attributes["axes"].ints();
-        axes             = std::vector<int64_t>(attr_axes.begin(), attr_axes.end());
-    }
+        auto constant_axes = parse_constant_axes(args, info);
 
-    bool noop_with_empty_axes = false;
-    if(contains(info.attributes, "noop_with_empty_axes"))
-    {
-        noop_with_empty_axes = static_cast<bool>(
-            parser.parse_value(info.attributes.at("noop_with_empty_axes")).at<int>());
-    }
+        int noop_with_empty_axes =
+            parse_attribute<int>("noop_with_empty_axes", parser, info).value_or(0);
 
-    // empty axes behavior
-    if(axes.empty())
-    {
-        if(noop_with_empty_axes)
+        int keep_dims = parse_attribute<int>("keepdims", parser, info).value_or(1);
+
+        std::vector<int64_t> all_axes(args.front()->get_shape().ndim());
+        std::iota(all_axes.begin(), all_axes.end(), 0);
+
+        // Handle axes attribute, constant input axes, and missing both attribute and input cases
+        if(constant_axes.has_value())
         {
-            return args.at(0);
+            if(noop_with_empty_axes != 0 and constant_axes->empty())
+                return args[0];
+
+            if(noop_with_empty_axes == 0 and constant_axes->empty())
+                constant_axes = all_axes;
+
+            auto reduce =
+                info.add_instruction(make_op(op_name, {{"axes", *constant_axes}}), args[0]);
+
+            if(keep_dims == 0)
+                return info.add_instruction(make_op("squeeze", {{"axes", *constant_axes}}), reduce);
+
+            return reduce;
+        }
+
+        // Handle variable input axes
+        if(keep_dims == 0)
+            MIGRAPHX_THROW("Keepdims not supported with runtime provided axes");
+
+        // Empty axes attribute indicates to the operator to look for axes in the inputs
+        auto reduce_op = make_op(op_name, {{"axes", {}}});
+
```

Comment:
Pardon me, I am a little confused here.. if the operator were to look for `axes` in its variable inputs (line #75) as per the comment above it, then how is line #77 below able to deduce that input `axis` is _empty_ -- that is the only case where the `noop_with_empty_axis` should activate, similar to line #55. Thanks.

Reply:
The op does not form any assumptions about input axes at this point. This is delegated to the [compute method](https://github.com/migraphx-benchmark/AMDMIGraphX/blob/5f7b702f88e13dc97ae8be108c11031046b5e1f9/src/include/migraphx/op/reduce_op.hpp#L231). There, if the variable axes are empty it will just return its input, if they're not it will proceed as usual(essentially it always presumes that noop_with_empty_axes is true)

It is worth nothing that following this code path is the only way in which an empty variable axes input can reach the operator. In the subsequent parsing code it can be seen that all branches will provide an axes input which is not empty, thus circumventing the check above for all cases except for the specific one for which it is relevant - noop_with_empty_axes is true, and the variable axes are empty.

Reply:
Perhaps I should have been clearer.. the compute method you refer to, its outcome is two ways: either the operator has axes as an input or it does not. But the line 77 should have a clause for checking the axes being empty. Right now it applies regardless.. Thanks.

Reply:
As created in line 77 the operator will behave as following:

1. [Check if its axes data member is empty](https://github.com/migraphx-benchmark/AMDMIGraphX/blob/5f7b702f88e13dc97ae8be108c11031046b5e1f9/src/include/migraphx/op/reduce_op.hpp#L228). Seeing that it is, it knows that axes must be in the input.
2. Knowing that axes is an input, [it proceeds to check if this input is empty](https://github.com/migraphx-benchmark/AMDMIGraphX/blob/5f7b702f88e13dc97ae8be108c11031046b5e1f9/src/include/migraphx/op/reduce_op.hpp#L231), and returns the data input if it is. This check has to be in the operator to cover off the case of a variable dynamic axes input.